alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EuclideanPhysics2DEnvironmentImpl.kt

```diff
@@ -0,0 +1,96 @@
+package it.unibo.alchemist.model.implementations.environments
+
+import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
+import it.unibo.alchemist.model.interfaces.Neighborhood
+import it.unibo.alchemist.model.interfaces.Node
+import it.unibo.alchemist.model.interfaces.environments.EuclideanPhysics2DEnvironment
+import it.unibo.alchemist.model.interfaces.geometry.GeometricShapeFactory
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.Euclidean2DShape
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.Euclidean2DShapeFactory
+
+/**
+ * Implementation of {@link EuclideanPhysics2DEnvironment}
+ */
+class EuclideanPhysics2DEnvironmentImpl<T> : Abstract2DEnvironment<T, Euclidean2DPosition>(), EuclideanPhysics2DEnvironment<T> {
+
+    override val shapeFactory: Euclidean2DShapeFactory = GeometricShapeFactory.getInstance()
+    private val defaultHeading = Euclidean2DPosition(0.0, 0.0)
+    private val nodeToHeading = mutableMapOf<Node<T>, Euclidean2DPosition>()
+    private var largestShapeDiameter: Double = 0.0
+
+    override fun getNodesWithin(shape: Euclidean2DShape): List<Node<T>> =
+        if (shape.diameter + largestShapeDiameter <= 0) emptyList()
+        else getNodesWithinRange(shape.centroid, (shape.diameter + largestShapeDiameter) / 2)
+            .filter { shape.intersects(getShape(it)) }
+
+    override fun getHeading(node: Node<T>) =
+        nodeToHeading.getOrPut(node, { defaultHeading })
+
+    override fun setHeading(node: Node<T>, direction: Euclidean2DPosition) {
+        val oldHeading = getHeading(node)
+        getPosition(node)?.also {
+            nodeToHeading[node] = direction
+            if (!canNodeFitPosition(node, it)) {
+                nodeToHeading[node] = oldHeading
+            }
+        }
+    }
+
+    override fun getShape(node: Node<T>): Euclidean2DShape =
+        shapeFactory.requireCompatible(node.shape)
+            .transformed {
+                origin(getPosition(node))
+                rotate(getHeading(node))
+            }
+
+    /**
+     * move this code in addNode when and if you merge this with AbstractEnvironment
```

Comment:
I could merge this with Abstract2DEnvironment and make it euclidean by default (should I keep it Java or go with Kotlin?).
If anyone wanted to make a non-euclidean implementation, he'd have to create another "Abstract2DNonEuclideanEnvironment".

Merging it with AbstractEnvironment should be possible but I'd have to add some generic parameters to it, then do the same thing with Abstract2DEnvironment (which could be euclidean by default to avoid the need to specify parameters in every subclass)

The third option would be to add the parameters to every abstract class and let the final ones define them.

 It makes no difference to me. If we are ok with 2D euclidean environments by default I could go with the first or second option without needing to touch the final classes

Reply:
I believe MapEnvironment, which uses GeoPosition, inherits from Abstract2DEnvironment. As such, it should not be euclidean by default.

I do not want to merge the two classes, but it's ok to move a method up one level in the hierarchy if it's generic enough. I thought you were referring to the method, not to the class.

Java <-> Kt: whatever is quicker and cleaner to you.

Reply:
I tried to move it up but it's not possible to do so without adding shapes support to every environment.
The node requires the environment to provide a shapeFactory to build the shape (even for the adimensional shape, because of its property _centroid_), the shapeFactory can only be built for a specific geometry, the method needs the shape of the node to calculate the diameter.

For the moment I've just written a proper documentation